File: ios/Classes/BarcodeReaderPlugin.swift
import Flutter
import UIKit

public class BarcodeReaderPlugin: NSObject, FlutterPlugin {
  private static var cameraViewInstance: CameraView?

  public static func register(with registrar: FlutterPluginRegistrar) {
    let channel = FlutterMethodChannel(name: "barcode_reader", binaryMessenger: registrar.messenger())

    let instance = BarcodeReaderPlugin()
    registrar.addMethodCallDelegate(instance, channel: channel)

    let factory = BarcodeReaderViewFactory(messenger: registrar.messenger())
    registrar.register(factory, withId: "barcode_reader_view")
  }

  public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
      switch call.method {

      case "toggleFlash":
          if let enabled = call.arguments as? Bool {
              if let cameraView = BarcodeReaderPlugin.cameraViewInstance {
                  cameraView.toggleFlash(enabled)
                  result(nil)
              } else {
                  result(FlutterError(code: "no_camera_view",
                                      message: "Camera view is not initialized.",
                                      details: nil))
              }
          } else {
              result(FlutterError(code: "invalid_argument",
                                  message: "Expected a boolean value for 'enabled'.",
                                  details: nil))
          }

      case "takePicture":
          if let view = BarcodeReaderPlugin.cameraViewInstance {
              view.takePicture { path in
                  if let path = path {
                      result(path)
                  } else {
                      result(FlutterError(code: "capture_failed", message: "Photo capture failed", details: nil))
                  }
              }
          } else {
              result(FlutterError(code: "no_camera_view", message: "Camera view is not initialized", details: nil))
          }

      case "pauseCamera":
          BarcodeReaderPlugin.cameraViewInstance?.pauseCamera()
          result(nil)

      case "resumeCamera":
          BarcodeReaderPlugin.cameraViewInstance?.resumeCamera()
          result(nil)

      default:
          result(FlutterMethodNotImplemented)
      }
  }

  static func setCameraView(_ view: CameraView) {
      cameraViewInstance = view
  }
}

File: android/src/main/kotlin/com/abomis/barcode_reader_plus/BarcodeReaderPlugin.kt
package com.abomis.barcode_reader_plus

import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel.MethodCallHandler

class BarcodeReaderPlugin : FlutterPlugin, MethodCallHandler {

  private lateinit var channel: MethodChannel
  private var viewFactory: BarcodeCameraViewFactory? = null

  override fun onAttachedToEngine(binding: FlutterPlugin.FlutterPluginBinding) {
    channel = MethodChannel(binding.binaryMessenger, "barcode_reader")
    channel.setMethodCallHandler(this)

    val factory = BarcodeCameraViewFactory(binding.binaryMessenger, channel)
    viewFactory = factory
    binding.platformViewRegistry.registerViewFactory("barcode_reader_view", factory)
  }

  override fun onMethodCall(call: MethodCall, result: MethodChannel.Result) {
    when (call.method) {
      "toggleFlash" -> {
        val enabled = call.arguments as? Boolean ?: false
        viewFactory?.toggleFlash(enabled)
        result.success(null)
      }
      "takePicture" -> {
        viewFactory?.takePicture { path ->
          if (path != null) {
            result.success(path)
          } else {
            result.error("capture_failed", "Photo capture failed", null)
          }
        }
      }
      else -> result.notImplemented()
    }
  }

  override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
    channel.setMethodCallHandler(null)
  }
}


